### PR TITLE
[suggest] change merge strategy： do not check write access if user in merge white list

### LINKED ISSUE #10935: [suggest] change merge strategy： do not check write access if user in merge white list
- Gitea version (or commit ref):  1.12.0+dev-82-g3723b0647
- Git version: 2.24
- Operating system: arch linux
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [x] No
  - [x] Not relevant
- Log gist:

## Description
https://github.com/go-gitea/gitea/blob/f685edf51073ac3be50ff00073a2432b4113bf82/services/pull/merge.go#L516
> before 
```go
// IsUserAllowedToMerge check if user is allowed to merge PR with given permissions and branch protections
func IsUserAllowedToMerge(pr *models.PullRequest, p models.Permission, user *models.User) (bool, error) {
	if !p.CanWrite(models.UnitTypeCode) {
		return false, nil
	}

	err := pr.LoadProtectedBranch()
	if err != nil {
		return false, err
	}

	if pr.ProtectedBranch == nil || pr.ProtectedBranch.IsUserMergeWhitelisted(user.ID) {
		return true, nil
	}

	return false, nil
}
```
> atter:
```go
// IsUserAllowedToMerge check if user is allowed to merge PR with given permissions and branch protections
func IsUserAllowedToMerge(pr *models.PullRequest, p models.Permission, user *models.User) (bool, error) {
	//if !p.CanWrite(models.UnitTypeCode) {
	//	return false, nil
	//}

	err := pr.LoadProtectedBranch()
	if err != nil {
		return false, err
	}

	if (p.CanWrite(models.UnitTypeCode) && pr.ProtectedBranch == nil) || pr.ProtectedBranch.IsUserMergeWhitelisted(user.ID) {
		return true, nil
	}

	return false, nil
}
```

#### remove Route verification
```go
m.Post("/merge", context.RepoMustNotBeArchived(), /*reqRepoPullsWriter,*/ bindIgnErr(auth.MergePullRequestForm{}), repo.MergePullRequest)
```
...

## Screenshots
image

### PR DESCRIPTION
**Pull request by everhopingandwaiting**

### CODE REVIEW

**Comment by 6543** [comment]
test's fail pleace correct tests :+1:

**Comment by everhopingandwaiting** [comment]
> test's fail pleace correct tests 👍

let me check check  🤔

File: services/pull/merge.go [hunk 1]
```diff
@@ -531,16 +531,13 @@ func IsSignedIfRequired(pr *models.PullRequest, doer *models.User) (bool, error)
...
-		return false, nil
-	}
 
 	err := pr.LoadProtectedBranch()
 	if err != nil {
 		return false, err
 	}
 
-	if pr.ProtectedBranch == nil || pr.ProtectedBranch.IsUserMergeWhitelisted(user.ID) {
+	if (p.CanWrite(models.UnitTypeCode) && pr.ProtectedBranch == nil) || pr.ProtectedBranch.IsUserMergeWhitelisted(user.ID) {
```
Comment (guillep2k) [inline services/pull/merge.go]:
```suggestion
	if (p.CanWrite(models.UnitTypeCode) && pr.ProtectedBranch == nil) || (pr.ProtectedBranch != nil && pr.ProtectedBranch.IsUserMergeWhitelisted(user.ID)) {
```

**Review by guillep2k — COMMENTED** [review]
In this PR you can check the tests link here:

image

If you follow the link, you will find that the error happens here:

image

Fix in my suggestion below.

File: services/pull/merge.go (hunk 1 above)
Comment (6543) [reply services/pull/merge.go]: thx for looking didn't had time until now to look at it
@everhopingandwaiting It should be fixed by apply guillep's code

Comment (everhopingandwaiting) [reply services/pull/merge.go]: Last night I did n’t check again for other reasons, thanks all project members 👍🍖

**Review by 6543 — APPROVED** [review]

**Review by guillep2k — APPROVED** [review]

**Review by 6543 — APPROVED** [review]
good chatch! for the API

**Comment by 6543** [comment]
ping lgtm

**Comment by 6543** [comment]
ping lgtm

**Review by lafriks — APPROVED** [review]

**Pull request merged**